src/persist-client/src/internal/compact.rs

```diff
@@ -369,84 +369,56 @@ where
             compaction_schema.id,
         );
 
+        let isolated_runtime = Arc::clone(&machine.isolated_runtime);
+        let machine_clone = machine.clone();
+        let metrics_clone = Arc::clone(&machine.applier.metrics);
         let compact_span = debug_span!("compact::consolidate");
         let res = tokio::time::timeout(
             timeout,
-            // Compaction is cpu intensive, so be polite and spawn it on the isolated runtime.
-            machine
-                .isolated_runtime
-                .spawn_named(
-                    || "persist::compact::consolidate",
-                    Self::compact(
-                        CompactConfig::new(&machine.applier.cfg, machine.shard_id()),
-                        Arc::clone(&machine.applier.state_versions.blob),
-                        Arc::clone(&metrics),
-                        Arc::clone(&machine.applier.shard_metrics),
-                        Arc::clone(&machine.isolated_runtime),
-                        req,
+            isolated_runtime.spawn_named(
+                || "persist::compact::consolidate",
+                async move {
+                    let stream = Self::compact_stream(
+                        CompactConfig::new(&machine_clone.applier.cfg, machine_clone.shard_id()),
+                        Arc::clone(&machine_clone.applier.state_versions.blob),
+                        Arc::clone(&metrics_clone),
+                        Arc::clone(&machine_clone.applier.shard_metrics),
+                        Arc::clone(&machine_clone.isolated_runtime),
+                        req.clone(),
                         compaction_schema,
+                    );
+                    let res = Self::compact_all(stream, req.clone()).await?;
+                    let maintenance = Self::apply(
+                        FueledMergeRes { output: res.output },
+                        &metrics_clone,
+                        &machine_clone,
                     )
-                    .instrument(compact_span),
-                )
-                .map_err(|e| anyhow!(e)),
-        )
-        .await;
+                    .await?;
 
-        let res = match res {
-            Ok(res) => res,
-            Err(err) => {
-                metrics.compaction.timed_out.inc();
-                Err(anyhow!(err))
-            }
-        };
+                    Ok::<_, anyhow::Error>(maintenance)
+                }
+                .instrument(compact_span),
+            ),
+        )
+        .await
+        .map_err(|e| {
+            metrics.compaction.timed_out.inc();
+            anyhow!(
+                "compaction timed out after {}s: {}",
+                timeout.as_secs_f64(),
+                e
+            )
+        })?
+        .map_err(|e| anyhow!(e))?;
```

Comment:
Here we're replacing the old `compact` method with a composition of the three new methods on the "main" codepath, but all the tests and admin tools still hit the old method. This means that we're keeping around some almost-dead code, and we're not giving the new stuff as much coverage as we otherwise would.

IMO it would be reasonable to either allow flagging between the old and new path, or to just implement `fn compact` with calls to the new methods...

Reply:
Good point, compact is reimplemented in terms of these new functions